### ISSUE TITLE
`collapsible_match` should not trigger for if statements that take ownership

### ISSUE THREAD

**Issue by elenaf9** [rust-lang/rust-clippy#16705]
### Summary

The lint triggers for if branches that should be collapsed into the match statement.
This is not possible if the if statement takes ownership of a pattern value.

### Lint Name

collapsible_match

### Reproducer

I tried this code:

```rust
fn is_foo(s: String) -> bool {
    drop(s);
    true
}

fn main() {
    let x: Option<String> = Some("Test".into());

    let bar = match x {
        Some(val) => {
            if is_foo(val) {
                return;
            } else {
                false
            }
        }
        _ => false,
    };
    println!("{bar}")
}

```

I saw this happen:

```sh
error: this `if` can be collapsed into the outer `match`
  --> src/main.rs:11:13
   |
11 | /             if is_foo(val) {
12 | |                 return;
13 | |             } else {
14 | |                 false
15 | |             }
   | |_____________^
   |
```

I expected to see this happen:

The lint should not trigger. The if branch can't be collapsed into the pattern, since `is_food` takes ownership of `val`.

If the if branch is collapse, it gives an error:
```sh
error[E0507]: cannot move out of `val` in pattern guard
  --> src/main.rs:10:29
   |
10 |         Some(val) if is_foo(val) => {
   |                             ^^^ move occurs because `val` has type `std::string::String`, which does not implement the `Copy` trait
```

### Version

```text
rustc 1.94.0 (4a4ef493e 2026-03-02)
binary: rustc
commit-hash: 4a4ef493e3a1488c6e321570238084b38948f6db
commit-date: 2026-03-02
host: x86_64-unknown-linux-gnu
release: 1.94.0
LLVM version: 21.1.8
```

**Comment by elenaf9** [comment]
The same applies when the if branch mutably borrows a pattern guard variable.
